test/adaptive_load/metrics_plugin_test.cc

```diff
@@ -0,0 +1,233 @@
+          /*requests_per_second=*/123,
+          /*actual_duration_seconds=*/10,
+          /*upstream_rq_total=*/2560,
+          /*response_count_2xx=*/320,
+          /*min_ns=*/400,
+          /*mean_ns=*/500,
+          /*max_ns=*/600,
+          /*pstdev_ns=*/11,
+      }));
+  EXPECT_THAT(plugin.GetMetricByName("nonexistent-metric-name").status().message(),
+              testing::HasSubstr("was not computed by the 'builtin' plugin"));
+}
+
+TEST(NighthawkStatsEmulatedMetricsPluginTest, DeterminesConcurrencyWithSingleWorker) {
+  NighthawkStatsEmulatedMetricsPlugin plugin =
+      NighthawkStatsEmulatedMetricsPlugin(MakeSimpleNighthawkOutput({
+          /*concurrency=*/"auto",
+          /*requests_per_second=*/123,
+          /*actual_duration_seconds=*/10,
+          /*upstream_rq_total=*/2560,
+          /*response_count_2xx=*/320,
+          /*min_ns=*/400,
+          /*mean_ns=*/500,
+          /*max_ns=*/600,
+          /*pstdev_ns=*/11,
+      }));
+  // |results| in output contains only "global" when there was 1 worker.
+  EXPECT_EQ(plugin.GetMetricByName("attempted-rps").value(), 123.0);
+}
+
+TEST(NighthawkStatsEmulatedMetricsPluginTest, DeterminesConcurrencyWithMultipleWorkers) {
+  nighthawk::client::Output nighthawk_output = MakeSimpleNighthawkOutput({
+      /*concurrency=*/"auto",
+      /*requests_per_second=*/123,
+      /*actual_duration_seconds=*/10,
+      /*upstream_rq_total=*/2560,
+      /*response_count_2xx=*/320,
+      /*min_ns=*/400,
+      /*mean_ns=*/500,
+      /*max_ns=*/600,
+      /*pstdev_ns=*/11,
+  });
+
+  // |results| when there are 2 workers contains 3 entries not necessarily in this order: "global"
+  // merged results, worker 1 individual results, worker 2 individual results.
+  *nighthawk_output.mutable_results()->Add() = nighthawk_output.results(0);
+  *nighthawk_output.mutable_results(1)->mutable_name() = "worker1";
+  *nighthawk_output.mutable_results()->Add() = nighthawk_output.results(0);
+  *nighthawk_output.mutable_results(2)->mutable_name() = "worker2";
+  ASSERT_EQ(nighthawk_output.results_size(), 3);
+
+  NighthawkStatsEmulatedMetricsPlugin plugin =
+      NighthawkStatsEmulatedMetricsPlugin(nighthawk_output);
+  EXPECT_EQ(plugin.GetMetricByName("attempted-rps").value(), 246.0);
+}
+
+TEST(NighthawkStatsEmulatedMetricsPluginTest, ReturnsCorrectSupportedMetricNames) {
+  NighthawkStatsEmulatedMetricsPlugin plugin = NighthawkStatsEmulatedMetricsPlugin({});
+  std::vector<std::string> supported_metrics = plugin.GetAllSupportedMetricNames();
+  EXPECT_THAT(supported_metrics,
+              ::testing::ElementsAre("attempted-rps", "achieved-rps", "send-rate", "success-rate",
+                                     "latency-ns-min", "latency-ns-mean", "latency-ns-max",
+                                     "latency-ns-mean-plus-1stdev", "latency-ns-mean-plus-2stdev",
+                                     "latency-ns-mean-plus-3stdev", "latency-ns-pstdev"));
+}
+
+} // namespace
+
```

Comment:
Would it make sense to also test that if you have a worker that has valid statistics and a worker that has erroneous statistics what would happen? Or multiple instances of global and such?

Reply:
For workers with erroneous statistics, and for duplicated "global," we're trusting the unit tests of the Nighthawk code that computes the "global" statistics.

I don't know for sure whether "global" can ever be missing (e.g. in the event of a Nighthawk Service error), but the implementation checks for it just in case.

Reply:
re: "I don't know for sure whether "global" can ever be missing (e.g. in the event of a Nighthawk Service error), but the implementation checks for it just in case."

I can imagine situations where Nighthawk's could successfully terminate without any latency measurements. E.g. lots of random jitter, a very short execution duration and a very low qps: chances are no requests will be initiated before execution terminates based on exceeding duration.  An edge case, but not unthinkable.

Reply:
Thanks, good to know.